es: support different elasticsearch connections per index

hunchback opened the pull request:
the following change adds the infra changes needed for pointing topology and flow each to a different ElasticSearch database (the extra change need to support this is extending the `skydive.yml` `analyzer.storage` to support flow connection and point to this path during `NewConnConfig` call (for flow). 

This would enable extra isolation between the two in productive environments.

For now default behavior (in which both flow and topology take connection configuration from `storage.elasticsearch` remains).

safchain commented:
ok to test

safchain commented:
rerun unit-tests

eranra commented on storage/elasticsearch/client.go:
```diff
@@ -83,34 +83,43 @@ type ElasticSearchClientInterface interface {
+	wg         sync.WaitGroup
+	name       string
+	mappings   Mappings
+	idxCfg     IndexConfig
```
Please do not use short names, change idx to index, it is confusing. so for this variable use indexConfig

hunchback replied:
And variables are typically shorter than structs, as a variable lives in a block or struct (constrained scope) while a struct lives within the module (much larger scope) ... thus in idiomatic golang (and other languages) variables tend to be shorter

hunchback replied:
ok will change idxCfg --> indexCfg

eranra commented on storage/elasticsearch/client.go:
```diff
@@ -83,34 +83,43 @@ type ElasticSearchClientInterface interface {
+}
+
+// ConnConfig describes the ElasticSearch configuration section
+type ConnConfig struct {
```
Change to ConnectionConfig

hunchback replied:
Using shot names is widly accepted in golang (as opposed for example to Java) srarting from reserved words eg func, chan, int; on to module names eg fmt; and variable names. 

Specifically conn as used widely in idiomatic golang code

eranra replied:
Yes, I see that in the code in many places (including the lib); so I'm ok with that change ... although very confusing to see conn and think what it stands for, connection is much more understandable.

eranra commented on storage/elasticsearch/client.go:
```diff
@@ -636,72 +642,86 @@ func (c *ElasticSearchClient) Stop() {
-func NewElasticSearchClient(url *url.URL, maxConns int, retrySeconds int, bulkMaxDocs int, bulkMaxDelay int) (*ElasticSearchClient, error) {
+func urlFromHost(host string) (*url.URL, error) {
+	urlStr := host
+	if !strings.HasPrefix(urlStr, "http://") && !strings.HasPrefix(urlStr, "https://") {
```
So host can be either hostname without http/s or full url ...  can you add explanation in default .yml configuration file

hunchback replied:
This issue is outside the scope of this PR as i did not change the format of the skydive.yml file.

Additionally omitting the http prefix is used elsewhere (although not consistently) in other places such as analyzers.

For now i rather this not be addressed

eranra replied:
K, you are correct that this needs attention also in other places, I submitting a new issue to track https://github.com/skydive-project/skydive/issues/964

hunchback replied:
ok

hunchback commented:
I think i would take #968 furthur, such that i can refer to memory, orientdb, elasticsearch (default entries defined under storage), but can also add more options such as elasticsearch2 (an alternative ES instace) and then i could move limits from analyzer.flow (or analyzer.storage) to storage.elasticsearch, storage.elasticsearch2, etc.

hunchback commented:
@nplanel / @safchain 

The proposed fix has some general code improvements:
 - moving all initialization to the c'tor (and out of `Start()` which is an interface method). 
 - adding of helper methods for reading from configuration (and eliminating duplicated reading of configuration parameters)
I could be merged regardless of the separate discussion on changes on skydive.yml storage configuration changes.

If nonetheless you prefer to also address changes in the configuration in the same go - then I would suggest the following:

```
analyzer:
  storage:
    flow:
      backend: elasticsearch.flow
      max_chan_size: 100000                        # previous max_flow_buffer_size
    topology:
      backend: elasticsearch.default

storage:
  elasticsearch:
    default:
      host: 127.0.0.1:9200
      bulk_max_conns: 10                                 # previously maxconns        
      bulk_retry_seconds: 60                             # previously retry
      bulk_insert_size: 100                                 # previously bulk_maxdocs, bulk_insert
      bulk_insert_seconds: 5                              # previously bulk_maxdelay, bulk_insert_deadline
      index_entries_limit: 0
      index_age_limit: 0
      indices_to_keep: 0
    flow:
      ...
  orientdb:
    ...
```

nplanel commented:
@eranra @hunchback @safchain 

1/ Would be better to not mix backend and indexes, there are two different things
2/ Let's keep the configuration simple and backward compatible
  - storage.elasticsearch define the default configuration (as today)
  - storage.elasticsearch.{index} define the {index} configuration (if exist, else fallback to default)
3/ analyzer flow_server bulk_insert and bulk_insert_deadline, use the configuration values according to the selected flow backend

```
analyzer:
  storage:
    flow:
      backend: elasticsearch
      max_flow_buffer_size: 100000
    topology:
      backend: elasticsearch

storage:
  elasticsearch:
    host: 127.0.0.1:9200
    maxconns: 10
    retry: 60
    bulk_maxdocs: 100
    bulk_maxdelay: 5
    index_entries_limit: 0
    index_age_limit: 0
    indices_to_keep: 0
    flow:                                                           # optional
      host: 127.0.1.1:9200
      ...
    topology:                                                    # optional
      host: 127.0.2.1:9200
      ...
    snort:                                                          # optional
      host: 127.0.3.1:9200
      ...
  orientdb:
    ...

```

hunchback commented:
@nplanel  (@safchain, @eranra) - proposed config seem fine, can this be handled in a separate PR?

nplanel commented:
I think the config need to be handled in the this PR as it's related to ES and indexes.

nplanel commented on storage/elasticsearch/elasticsearch_test.go:
```diff
@@ -27,121 +27,122 @@ const testMapping = `
 func TestElasticsearchShouldRollByCount(t *testing.T) {
-	limits := ElasticLimits{}
-	limits.EntriesLimit = 5
+	indexCfg := IndexConfig{}
```
please don't rename if it's not improving IndexConfig is exactly the same thing as ElasticLimits
let's focus on the patch behaviour, the review will be more readable

hunchback commented:
Ok

On Sat, Apr 21, 2018, 12:53 Nicolas PLANEL <notifications@github.com> wrote:

> I think the config need to be handled in the this PR as it's related to ES
> and indexes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/962#issuecomment-383282160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACogI-WBvDQkhHl90PJrmBaN6ic0B1idks5tqwG2gaJpZM4TYkOE>
> .
>

hunchback commented:
@nplanel - I have just uploaded an updated proposal - to handle default values I and added a notation with '?' such as `storage.elasticsearch?flow.host` which means check if `storage.elasticsearch.flow.,host` is set if not then fallback to `storage.elasticsearch.host`

nplanel commented:
@hunchback 
why adding "?" ?
why not try to resolve the key like

```
host := ""
if host = cfg.GetString("storage.elasticsearch.flow.host"); host == "" {
    host = cfg.GetString("storage.elasticsearch.host")
}
```

nplanel commented on config/config.go:
```diff
@@ -89,12 +104,10 @@ func init() {
-	cfg.SetDefault("analyzer.storage.index_entries_limit", 0)
-	cfg.SetDefault("analyzer.storage.indices_to_keep", 0)
-	cfg.SetDefault("analyzer.storage.max_flow_buffer_size", 100000)
+	cfg.SetDefault("analyzer.storage.flow.bulk_insert", 100)
```
please take a look at #968 
graph.backend => analyzer.topology.backend
analyzer.storage => analyzer.flow

analyzer.flow.{bulk_insert,bulk_insert_deadline,max_flow_buffer_size} must be removed from the configuration and deduced/resolved from the used backend (analyzer.flow.backend)

hunchback replied:
ok, note that now:

analyzer.flow.{bulk_insert,bulk_insert_deadline,max_flow_buffer_size} ==> analyzer.storage.flow.{bulk_insert,bulk_insert_deadline,max_flow_buffer_size}

hunchback commented:
I am using the '?' notation so as to keep the calling code simple.

Additionaly, by adding the '?' notation i was able to extend all existing
config.GetXXX(key string) functions with only minimal code changes (and
minimal added complexity)

On Mon, Apr 23, 2018, 19:47 Nicolas PLANEL <notifications@github.com> wrote:

> @hunchback <https://github.com/hunchback>
> why adding "?" ?
> why not try to resolve the key like
>
> host := ""
> if host = cfg.GetString("storage.elasticsearch.flow.host"); host == "" {
>     host = cfg.GetString("storage.elasticsearch.host")
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/962#issuecomment-383643815>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACogI-jEHMsWMl8VLwwlsU4SCZs2yyELks5trgWDgaJpZM4TYkOE>
> .
>

hunchback commented:
I will fix the handling of deprecated keys according to example #968

On Mon, Apr 23, 2018, 20:10 Nicolas PLANEL <notifications@github.com> wrote:

> *@nplanel* commented on this pull request.
> ------------------------------
>
> In config/config.go
> <https://github.com/skydive-project/skydive/pull/962#discussion_r183470284>
> :
>
> > @@ -89,12 +104,10 @@ func init() {
>  	cfg.SetDefault("analyzer.packet_injection_enabled", true)
>  	cfg.SetDefault("analyzer.replication.debug", false)
>  	cfg.SetDefault("analyzer.ssh_enabled", false)
> -	cfg.SetDefault("analyzer.storage.bulk_insert", 100)
> -	cfg.SetDefault("analyzer.storage.bulk_insert_deadline", 5)
> -	cfg.SetDefault("analyzer.storage.index_age_limit", 0)
> -	cfg.SetDefault("analyzer.storage.index_entries_limit", 0)
> -	cfg.SetDefault("analyzer.storage.indices_to_keep", 0)
> -	cfg.SetDefault("analyzer.storage.max_flow_buffer_size", 100000)
> +	cfg.SetDefault("analyzer.storage.flow.bulk_insert", 100)
>
> please take a look at #968
> <https://github.com/skydive-project/skydive/pull/968>
> graph.backend => analyzer.topology.backend
> analyzer.storage => analyzer.flow
>
> analyzer.flow.{bulk_insert,bulk_insert_deadline,max_flow_buffer_size} must
> be removed from the configuration and deduced/resolved from the used
> backend (analyzer.flow.backend)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/962#pullrequestreview-114474467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACogI6XPUY1rB4fKFSPFFMWpHY5eLU6Lks5trgrxgaJpZM4TYkOE>
> .
>

hunchback commented:
I am using the '?' notation so as to keep the calling code simple.

Additionally, by adding the '?' notation i was able to extend all existing
config.GetXXX(key string) functions with only minimal code changes (and
minimal added complexity).

Compare: 

```
retry := 0
if val := cfg.Get("storage.elasticsearch.flow.retry"); val != nil {
    retry = val.(int)
} else {
    retry = cfg.GetInt("storage.elasticsearch.retry")
}
```
 To

```
retry := cfg.GetInt("storage.elasticsearch?flow.retry")
```

In our case things are ever more acute as we have a helper function handling the calling of configuration which is also simplified to:

In your case it would be:

```
cfg := es.NewConfig("storage.elasticsearch.flow", "storage.elasticsearch")
```

Simplified to:

```
cfg := es.NewConfig("storage.elasticsearch?flow")
```

nplanel commented:
@hunchback the only call you need to do is 

```
limits := elasticsearch.NewElasticLimitsFromConfig("storage.elasticsearch", "flow")

or even you can simplify

limits := elasticsearch.NewElasticLimitsFromConfig("flow")
limits := elasticsearch.NewElasticLimitsFromConfig("topology")
...
limits := elasticsearch.NewElasticLimitsFromConfig()
```

hunchback commented:
@nplanel - I agree there is little difference between:

```
limits := elasticsearch.NewElasticLimitsFromConfig("storage.elasticsearch", "flow")
```

And:

```
limits := elasticsearch.NewElasticLimitsFromConfig("storage.elasticsearch?flow")
```

But this is a negligible aspect - the main issue is `config.GetInt("storage.elasticsearch?flow.retry")` and friends if not handled seamlessly by `config` then we have a 7-8 way code duplication, and this will only get worse we we use the default/fallback trick elsewhere in the code (in storage.orientdb for example).

hunchback commented:
Tests failed due to:

```
2018-04-24T07:59:03.582Z	DEBUG	helper/helper.go:135 ExecCmds	skydive-slave.rdocloud: Executing command [lxc launch images:fedora/27 test-skydive-lxd-simple]
2018-04-24T07:59:03.628Z	DEBUG	helper/helper.go:138 ExecCmds	skydive-slave.rdocloud: Command returned Connection refused; is LXD running?

```

@nplanel  - can you check was is the matter?

nplanel commented:
run skydive-functional-tests-backend-elasticsearch

lebauce commented:
I'm not a fan of the "?" notation and also prefer ```
cfg := es.NewConfig("storage.elasticsearch.flow", "storage.elasticsearch")
```

One way would be to make the GetString, GetInt take a ```...string``` argument and returns the first that is set, but I'm not sure it's worth it.

There are changes in your PR related to the proposal, but it seems the PR is not rebased of it, that makes it a bit hard to understand.

hunchback commented:
I thought about extending config.GetInt() and friends ... but this has two
shortcomings:

1.  Usage is more comborsum, eg: GetInt("storage.elasticsearch.flow.retry",
"storage.elasticsearch.retry") compared to
GetInt("storage.elasticsearch?flow.retry")

2. Extending interfaces will cascade down to extending each functions'
implementation ...

So i don't see any advantage of this over the '?' notation .. and would
much prefer we stay with my current implementation. Also note this notation
is not exposed to users - only used from within the code.

As to your comment on rebase - please elaborate
On Apr 24, 2018 6:34 PM, "Sylvain Baubeau" <notifications@github.com> wrote:

> I'm not a fan of the "?" notation and also prefer ```
> cfg := es.NewConfig("storage.elasticsearch.flow", "storage.elasticsearch")
>
>
> One way would be to make the GetString, GetInt take a ```...string``` argument and returns the first that is set, but I'm not sure it's worth it.
>
> There are changes in your PR related to the proposal, but it seems the PR is not rebased of it, that makes it a bit hard to understand.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/962#issuecomment-383977471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACogI_TqEH-22uFlgJoEFFhfnyhuzDQeks5tr0YRgaJpZM4TYkOE>
> .
>

lebauce commented:
It's not clear what `GetInt("storage.elasticsearch?flow.retry")` (at least for me). And it only allows simple use cases such as "look inside the child then in the parent if you didn't find it".

GetInt("storage.elasticsearch.flow.retry", "storage.elasticsearch.retry") seems a bit clearer on what it does and allows the keys to be completely different.
That being said, maybe we should the KISS approach and go with:

```
config := cfg.Get("storage.elasticsearch.flow")
if config != nil {
    ...
} else {
    config = cfg.Get("storage.elasticsearch")
}
```
as it's only used once for now. We'll still be able to generalize that if we use it for something else. What do you think ?

hunchback commented:
@lebauce - the snipped you provided will fail generally, as we want the ability per each parameter to get the value first from "storage.elasticseach.flow" and fallback to "storage.elasticsearch". For example:

```
storage:
  elasticsearch:
    retry: 10
    flow:
    # no retry value
```
In your snippet you will incorrectly return `0` rather than `10`

In effect extending the key syntax provides a means of extending capabilities while only having to "fix" one function - `realKey()` - so this is the generic way forward.

hunchback commented:
@lebauce - I am not very happy with the last commit (below), I would much rather drop it ...

https://github.com/skydive-project/skydive/pull/962/commits/dc3f024832f2796a8651daf0f72e01df464c8215

nplanel commented on config/config_test.go:
```diff
@@ -76,3 +77,60 @@ openstack:
+	if expected != actual {
+		t.Fatal(fmt.Sprintf("Config 'read_from_parent': expected %d; actual %d", expected, actual))
+	}
+}
```
please test GetInt/GetString("a", "b", "c") should print an error message (see capturer TestRelocation in config/config_test.go

hunchback replied:
I just test the case of no keys - only testing GetInt (as GetString, GetBool, etc, etc) are all using the same underlying `readKey()` func.

hunchback replied:
ok

nplanel commented on storage/elasticsearch/client.go:
```diff
@@ -47,24 +47,42 @@ const indexVersion = 11
-	limits := ElasticLimits{}
-	limits.EntriesLimit = config.GetInt(path + ".index_entries_limit")
-	limits.AgeLimit = 0
+func NewConfig(name string) Config {
```
NewConfig(name ...string)
if len(name) > 0 ...

to avoid call like NewConfig("")

hunchback replied:
ok

nplanel commented on storage/elasticsearch/client.go:
```diff
@@ -47,24 +47,42 @@ const indexVersion = 11
+		first += "." + name
+	}
+
+	cfg.ElasticHost = config.GetString(first+".host", second+".host")
```
as everything is almost constant and static why not writing

se := "storage.elasticsearch"
cfg.ElasticHost = config.GetString(se + ".host", se+"."+name+".host" )

hunchback replied:
this will break as `name` may not be "" and in that case what you proposed will evaluate to: 
```
GetString("storage.elasticsearch.host", "storage.elasticseach..host")
```

hunchback commented on config/config.go:
```diff
@@ -350,7 +362,19 @@ func IsTLSenabled() bool {
 }
 
-func realKey(key string) string {
+func expandKey(keys ...string) string {
```
ok

hunchback commented on config/config.go:
```diff
@@ -350,7 +362,19 @@ func IsTLSenabled() bool {
+			break
+		}
+	}
+	return k
```
ok

hunchback commented on config/config.go:
```diff
@@ -350,7 +362,19 @@ func IsTLSenabled() bool {
+	var k string
+	for _, k = range keys {
+		if cfg.IsSet(k) {
+			break
```
ok

nplanel commented on config/config.go:
```diff
@@ -44,6 +44,21 @@ var ErrNoAnalyzerSpecified = errors.New("No analyzer specified in the configurat
 var (
 	cfg           *viper.Viper
 	relocationMap = map[string][]string{
+		"analyzer.storage.flow.backend": {
```
let's focus this pr only on connection per index
PR #968 will address new config namespace

hunchback replied:
ok, i have take code related to relocation, as it is more complete that #968 - I have created a new PR from branch `storage-relocate-backend` - @safchain - please suggest how/who will handle.

hunchback commented on topology/graph/elasticsearch.go:
```diff
@@ -615,48 +613,25 @@ func (b *ElasticSearchBackend) IsHistorySupported() bool {
-
 // NewElasticSearchBackendFromConfig creates a new graph backend based on configuration file parameters
 func NewElasticSearchBackendFromConfig() (*ElasticSearchBackend, error) {
-	elasticHost := config.GetString("storage.elasticsearch.host")
```
This is now resolved by `elasticsearch.NewConfig("topology")` together with the rest of the sotrage parameters.

hunchback commented on config/config_test.go:
```diff
@@ -76,3 +77,82 @@ openstack:
+
+	stderr := capturer.CaptureStderr(func() {
+		GetInt()
+	})
```
ok

hunchback commented:
Failed due to:

```
--- FAIL: TestPacketInjectionCapture (136.31s)
	selenium_test.go:177: By default 8 rows should be return, but got: 3
FAIL
```

Which does not appear related to my fix

lebauce commented:
run skydive-selenium-tests

hunchback commented:
can this be merged ?

nplanel commented on analyzer/flow_server.go:
```diff
@@ -114,7 +114,7 @@ func (c *FlowServerWebSocketConn) Serve(ch chan *flow.Flow, quit chan struct{},
 // NewFlowServerWebSocketConn returns a new WebSocket flow server
 func NewFlowServerWebSocketConn(server *shttp.Server) (*FlowServerWebSocketConn, error) {
-	flowsMax := config.GetConfig().GetInt("analyzer.storage.max_flow_buffer_size")
+	flowsMax := config.GetConfig().GetInt("analyzer.storage.flow.max_flow_buffer_size")
```
this out of scope of this PR

nplanel commented on config/config_test.go:
```diff
@@ -76,3 +77,101 @@ openstack:
+
+	cfg.ReadConfig(bytes.NewBuffer(yamlv1))
+
+	val := GetInt("a", "b", "c")
```
this test should report an warning

hunchback replied:
see above

hunchback commented on config/config.go:
```diff
@@ -350,7 +357,23 @@ func IsTLSenabled() bool {
+			return k
+		}
+	}
+	return k
```
But this will change the current behavior which does not report attempting to read non-defined keys ... for example currently if run `GetInt("a")` no warning will be emitted. 

If you only want `GetInt("a", "b")` to emit a warming but not `GetInt("a")` then this would make  things inconsistent.

If you want both `GetInt("a", "b")` and `GetInt("a")` to emit warnings if keys not set - they this is a change that should be done outside the scope of this PR (as it will require changing at least one non-related test in config_test.go

lebauce replied:
Ok , then let's remove the warning (the current behavior) and return ""

hunchback replied:
ok, so I am returning "" (other than that leaving "as-is")

hunchback commented:
Selenium is failing:

```
5b2ce5b57e9b: Pull complete
Digest: sha256:df5655cafb23744708d85653156b295979da25eb11eb85b4498219ed38212a99
Status: Downloaded newer image for docker.io/elgalu/selenium:latest
...
...
2018-05-02T16:38:23.576Z	INFO	ovs/ovsdb.go:194 (*OvsMonitor).interfaceUpdated	jenkins-slave-5.rdocloud: Interface "vm2-eth0(ac79407d-c37b-4e4f-b0c6-05519a89c807)" updated
panic: test timed out after 10m0s
```

hunchback commented:
@nplanel - if key is not deprecated and a "" is returned by isDepKeySet()
then firstKetSet() checks if key is set - if so then *key* is returned by
firstKeySet()

On Wed, May 2, 2018, 21:57 Nicolas PLANEL <notifications@github.com> wrote:

> *@nplanel* commented on this pull request.
> ------------------------------
>
> In config/config.go
> <https://github.com/skydive-project/skydive/pull/962#discussion_r185603606>
> :
>
> >  	// check is there is a deprecated key that can be used
>  	depKeys, found := relocationMap[key]
>  	if !found {
> -		return key
> +		return ""
>
> you break all good keys not relocated ....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/962#pullrequestreview-117025954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACogI51SChSsdgygsCW-N7VvDuc-vd5cks5tugGzgaJpZM4TYkOE>
> .
>

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-selenium-tests

safchain commented on etc/skydive.yml.default:
```diff
@@ -262,6 +262,11 @@ storage:
     # A value of 0 specifies no limit (i.e. indices will never be deleted)
     # indices_to_keep: 0
+    
+    # NOTE: you can override with specific values any of the above values
```
I'm really not a big fan of overriding flow, topology, etc here. What about named backend/connection that you can pick to use in the backend parameter, like

```
storage:
   es_1:
      type: elasticsearch
      maxconns: 200
   es_2:
      type: elasticsearch
      maxconns: 400

analyzer:
   flow:
      backend: es_2
   topology:
      backend: es_1
```

hunchback replied:
This raise two separate issues:

(1) should we use defaults? (such as storage.elasticsearch.flow and fallbacking to storage.elasticsearch) - as this significantly reduces the size of configuration I think this makes a lot of sense.

(2) to we use labels for backend, so as backend can also be memory or orientdb you would rather have to support values such as:

```
analyzer:
  flow:
    # backend: memory
    # backend: orentdb
    # backend: elasticsearch
    # backend: elasticsearch:es_2
```
But I am not sure this extra flexibility is required ... (as compared with defining backend:elasticsearch and having the logic of where to look for configuration coded within the code)

hunchback commented:
@nplanel / @safchain / @lebauce  - please review https://github.com/skydive-project/skydive/pull/962/commits/f9e3e066d1cd2d44a997f0ba35825b52d3091797
